### PR TITLE
#8268 Move Datadog Browser SDK logging to offscreen doc

### LINKED ISSUE #8268: Datadog error reporting does not work in background worker on MV3 builds
Looks like the problem is Datadog doesn't support service workers:
- https://github.com/DataDog/browser-sdk/issues/2198
- https://github.com/DataDog/browser-sdk/issues/432

Looks like there are a couple of potential workarounds:
- Use the Datadog API directly
https://eng.grip.security/enabling-chrome-extension-logging-with-datadog-a-journey-of-trial-and-error?utm_source=github
- Implement a hacky polyfill:
https://github.com/DataDog/browser-sdk/issues/432#issuecomment-1309227575

From @mnholtz: 
> I recommend that we set some kind of alerting immediately to flag when the approach breaks, either in CI or via a DD monitor. Preferably something in CI

### PR DESCRIPTION
**Pull request by mnholtz**
## What does this PR do?

- Closes #8268 
- The Datadog Browser SDK that we use to report error logs in the background is not compatible with service workers
- This is a workaround to moving that logic to an offscreen doc in order to get around those limitations. 
- The background script/service worker will relay a message to the offscreen doc once it's created to record the error using the SDK instead.

## Discussion

- It is possible that this change will get rejected by CWS due to the `Reason` passed to `offscreen.createDocument`, as no available reasons really fit our use-case. In that case, we'll likely have to resort to https://github.com/pixiebrix/pixiebrix-extension/pull/8269
- Let's hope that is not the case, because IMO this is the best approach 🤞 

## Demo

- See [this Datadog error](https://app.datadoghq.com/logs?query=-service%3A%28django%20OR%20www%29%20message%3ATest%20&cols=host%2Cservice&event=AgAAAY7uAF2rVdXMpQAAAAAAAAAYAAAAAEFZN3VBTlFZQUFBcGQ4dEJhT2dLWmdBQgAAACQAAAAAMDE4ZWVlMDEtMGZmYy00MjljLTlkNjEtZTQxNDQ5MDcxZDhh&fromUser=true&index=%2A&messageDisplay=inline&refresh_mode=sliding&storage=hot&stream_sort=desc&viz=stream&from_ts=1713387347015&to_ts=1713388247015&live=true) I sent from my local build
- Also see e2e test that demonstrates that error logging works in both MV2/MV3, edge, and chrome

## Checklist

- [x] Add jest or playwright tests and/or storybook stories
- [x] Designate a primary reviewer @grahamlangford

### CODE REVIEW

File: src/tinyPages/offscreen.ts
Comment (mnholtz) [inline src/tinyPages/offscreen.ts:1]: Other opportunity I see here is potentially using web-ext messenger for messaging. I went with this approach for ease making the POC

File: scripts/__snapshots__/manifest.test.js.snap [hunk 1]
```diff
@@ -107,6 +107,7 @@ exports[`customizeManifest beta 1`] = `
     "webNavigation",
     "contextMenus",
     "alarms",
+    "offscreen",
```
Comment (grahamlangford) [inline scripts/__snapshots__/manifest.test.js.snap:110]: Have we validated if this will disable the extension using the validation tool?

File: src/telemetry/logging.ts [hunk 2]
```diff
@@ -367,29 +400,28 @@ export async function reportToApplicationErrorTelemetry(
...
-    // Error reported not initialized
-    return;
-  }
-
-  const details = await selectExtraContext(error);
+  await setupOffscreenDocument("offscreen.html");
+
+  const { version_name: versionName } = chrome.runtime.getManifest();
+  const telemetryUser = await mapAppUserToTelemetryUser(await readAuthData());
+  const extraContext = await selectExtraContext(error);
```
Comment (grahamlangford) [inline src/telemetry/logging.ts:407]: Consider using `await Promise.all()` here

File: scripts/__snapshots__/manifest.test.js.snap (hunk 1 above)
Comment (mnholtz) [reply scripts/__snapshots__/manifest.test.js.snap:110]: Not yet - will add this to the list of TODOs

Comment (mnholtz) [reply scripts/__snapshots__/manifest.test.js.snap:110]: Confirmed that changing permissions seems to be ok with the validation tool

File: src/layout/ErrorDisplay.tsx
Comment (mnholtz) [inline src/layout/ErrorDisplay.tsx:1]: Sort-of related - I was using the Mods page to trigger an error for testing purposes. I noticed that we've been swallowing errors on that page from a logging perspective.

File: end-to-end-tests/tests/telemetry/errors.spec.ts [hunk 3]
```diff
@@ -31,6 +31,7 @@ test("can report application error to telemetry service", async ({
   await page.goto(getBaseExtensionConsoleUrl(extensionId));
   await expect(page.getByText("An error occurred")).toBeVisible();
 
+  // Due to limitations with the Datadog SDK, we need to report errors via an offscreen document.
```
Comment (twschiller) [inline end-to-end-tests/tests/telemetry/errors.spec.ts:34]: NIT: link to the Datadog server worker issue. The "limitations" are that SDK requires a window/document context

File: src/telemetry/logging.ts [hunk 4]
```diff
@@ -321,6 +324,36 @@ const THROTTLE_AXIOS_SERVER_ERROR_STATUS_CODES = new Set([502, 503, 504]);
...
+    // @ts-expect-error -- the type seems to be wrong here?
+    contextTypes: ["OFFSCREEN_DOCUMENT"],
+    documentUrls: [offscreenUrl],
+  });
+
+  if (existingContexts.length > 0) {
+    return;
+  }
+
+  if (creating) {
```
[the remark this answers is not in the document]
Comment (mnholtz) [reply src/telemetry/logging.ts]: How would you suggest I use memoizeUntilSettled to block returning until the promise resolves?

File: src/tinyPages/offscreen.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+  return (
+    "target" in message &&
+    message.target === "offscreen-doc" &&
+    "type" in message &&
+    message.type === "record-error"
+  );
+}
+
+async function handleMessages(message: unknown) {
```
Comment (mnholtz) [inline src/tinyPages/offscreen.ts:89]: Same logic that was in logging.ts, moved here

File: src/telemetry/logging.ts [hunk 6]
```diff
@@ -367,29 +373,37 @@ export async function reportToApplicationErrorTelemetry(
...
+
+  const recordErrorMessage: RecordErrorMessage = {
+    type: "record-error",
+    target: "offscreen-doc",
+    data: {
+      error,
+      errorMessage,
+      errorReporterInitInfo: {
+        // It should never happen that versionName is undefined, but handle undefined just in case
+        versionName: versionName ?? "",
```
Comment (mnholtz) [inline src/telemetry/logging.ts:396]: @twschiller @grahamlangford Any thoughts on what we might want to default here?

**Comment by mnholtz** [comment]
There is something unusual going on with error displays in MV2. E2E test is passing on MV3, but in MV2 I can't get the ErrorDisplay view to get displayed on the Mods Page? https://github.com/pixiebrix/pixiebrix-extension/blob/7a13481ee8d736090b7856205530817b7e52408a/src/extensionConsole/pages/mods/ModsPage.tsx#L39-L39

File: src/tinyPages/offscreen.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// import { selectExtraContext } from "@/data/service/errorService";
```
Comment (grahamlangford) [inline src/tinyPages/offscreen.ts:18]:
```suggestion
```

File: src/tinyPages/offscreen.ts [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+  if (existingContexts.length > 0) {
+    return;
+  }
+
+  if (createOffscreenDocumentPromise == null) {
+    createOffscreenDocumentPromise = chrome.offscreen.createDocument({
+      url: "offscreen.html",
+      // Our reason for creating an offscreen document does not fit nicely into options offered by the Chrome API, which
+      // is error telemetry as of 1.8.12 (we use this as a workaround for Datadog SDK service worker limitations).
```
Comment (grahamlangford) [inline src/tinyPages/offscreen.ts:62]:
```suggestion
      // is error telemetry as of 1.8.13 (we use this as a workaround for Datadog SDK service worker limitations).
```

File: src/telemetry/logging.ts (hunk 6 above)
Comment (grahamlangford) [reply src/telemetry/logging.ts:396]: As you commented, this is so unlikely to happen, I'm not worried about it

**Review by grahamlangford — APPROVED** [review]

File: end-to-end-tests/tests/telemetry/errors.spec.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        value
+          .url()
+          .startsWith(`chrome-extension://${extensionId}/offscreen.html`),
+      );
+
+    expect(offscreenPage.url()).toBeDefined();
+  }).toPass({ timeout: 5000 });
+
+  // TODO: due to Datadog SDK implementation, it will take ~30 seconds for the
+  //  request to be sent. We should figure out a way to induce the request to be sent sooner.
```
Comment (fungairino) [inline end-to-end-tests/tests/telemetry/errors.spec.ts:51]: Did you submit a support ticket to datadog about this?

Comment (mnholtz) [reply end-to-end-tests/tests/telemetry/errors.spec.ts:56]: Not for this particular issue, yet. I had another support ticket thread open with them about using their API
